integrate WDB.1 and WDB.2 with generic database server

jni2000 opened the pull request:
Please review the integration of WDB.1 and WDB.2 with the generic database server.

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_1', 'pal_db_record.json')
```
changed as suggested.

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -125,21 +132,34 @@ def test_WINNF_FT_S_WDB_1(self, config_filename):
-    # Step 3: Create PAL database with a PAL record containg the PAL ID 'P'.
+    # Step 3: Create PAL database with a PAL record containing the PAL ID 'P'.
+    pal_database = DatabaseServer('PAL Database',config['palDatabaseConfig']['hostName'],
+                                  config['palDatabaseConfig']['port'], authorization=True)
```
removed

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_2', 'CPI_Database-Public.csv')
```
changed as suggested

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
     # Step 2: Create a CPI database which includes CPI user with credentials
     # matching those used to create the request in Step 1.
+    cpi_database = DatabaseServer('CPI Database',config['cpiDatabaseConfig']['hostName'],
+                                  config['cpiDatabaseConfig']['port'],authorization=True)
```
removed

jni2000 commented on src/harness/testcases/testdata/wdb_1/pal_db_record.json:
```diff
@@ -0,0 +1,28 @@
+    "registrationDate": "2017-06-01"
+  },
+  "license": {
+    "licenseId": "fccId123456789 - 1",
```
removed.

jni2000 commented on src/harness/testcases/testdata/wdb_1/pal_db_record.json:
```diff
@@ -0,0 +1,28 @@
+    "licenseDate": "2017-07-01T00:00:00Z",
+    "licenseExpiration": "2020-07-01T00:00:00Z",
+    "licenseAreaIdentifier": "CensusTract123456789,Long Name",
+    "licenseAreaExtent": "zone\/census_tract\/census\/2010\/123456789",
```
removed

jni2000 commented on src/harness/testcases/testdata/wdb_1/pal_db_record.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "palId": "pal\/10-2017\/123456789\/A",
```
removed

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+                                config['cpiDatabaseConfig']['filePath'])
+
+    # Inject the CPI database URL into the SAS UUT.
+    self._sas_admin.InjectDatabaseUrl(config['cpiDatabaseConfig']['fileUrl'])
```
fixed as suggested

kate-harrison replied:
Confirmed, you should inject the URL to the index here.

jni2000 commented on src/harness/testcases/testdata/wdb_1/pal_db_record.json:
```diff
@@ -0,0 +1,28 @@
+  "userId": "cbsdOp1",
+  "registrationInformation": {
+    "frn": "123456-789",
+    "registrationDate": "2017-06-01"
```
Can you please send me a copy of the PAL record format you referred to? The latest we got from Kate is as below, which is what we used.

{ 
“palId” : “pal/10-2017/123456789/A”, 
“licensee” : “CBSD Operator No.1”, 
“userId” : “cbsdOp1”, 
“registrationInformation” : { 
“frn” : “123456-789”, 
“registrationDate”: “2017-06-01”, 
}, 
“license”: { 
“licenseId”: “fccId123456789-1”, 
“licenseDate”: “2017-07-01T00:00:00Z”, 
“licenseExpiration”: “2020-07-01T00:00:00Z”, 
“licenseAreaIdentifier”: “Census Tract 123456789, Long Name”, 
“licenseAreaExtent”: “zone/census_tract/census/2010/123456789”, 
“licenseFrequencyChannelId”: “1” 
}, 
“licenseStatus”: “VALID”, 
“channelAssignment”: { 
“primaryAssignment”: { 
“lowFrequency”: 3550000000, 
“highFrequency”: 3560000000 
}, 
“secondaryAssignment”: { 
“lowFrequency”: 3580000000, 
“highFrequency”: 3590000000 
} 
} 
}

kate-harrison replied:
https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf

jni2000 commented on src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv:
```diff
@@ -0,0 +1,3 @@
+cpiId,status,publicKeyIdentifier 
+ABCdef123,ACTIVE,
```
removed the invalid line in the original file.

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
+    pal_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
```
fixed using the format of FDB.3

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
+    cpi_database_config = {
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
```
fixed using the format of FDB.3

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
+        'hostName': 'localhost',
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_2', 'CPI_Database-Public.csv')
```
I do not recall that we will have multiple CPI database files. Can you please point me to the document or reference ? Thanks.

TaranFurnell replied:
The test CPI database has to host both the database index file (the one that the SAS is informed about), as well as the individual private keys for each CPI (which it gets the url for from the database).

After a bit more discussion with Kate we have the following suggested solution:

As part of the CPI database config have a field `cpis` set similar to the following
```python
'cpis': [
    {'id': 'abcd', 'status': 'ACTIVE', 'public_key_file': 'some/file/path'},
    {'id': 'defg', 'status': 'ACTIVE', 'public_key_file': 'some/other/file/path'}
]
```

and the `indexUrl` as already asked for.

Then in the test case itself use this information to create the index file, each line looks something like:
```python
cpi['id'] + ',' + cpi['status'] + ',/' + cpi['public_key_file']
```

and set the contents of the database server to be:
```python
files = {('/'+file['public_key_file']): file['public_key_file'] for file in config[cpis]}
files[config['indexUrl']] = # whatever file path the index file was saved at.
cpi_database.setFilesToServe(files)
```

kate-harrison replied:
I think you missed part of this: the index file can be written in the test case itself, and does not need to be committed to GitHub.

janetliuboston replied:
Fixed as per suggestion.
To generate CSV file, I have used the first instance of cpis information like below 
config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'] 
Please let me know this is correct way of doing it?

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +231,35 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+
+    # Set file path.
+    cpi_database.setFileToServe(config['cpiDatabaseConfig']['fileUrl'],
+                                config['cpiDatabaseConfig']['filePath'])
```
fixed as suggested.

jni2000 commented on src/harness/testcases/testdata/wdb_2/CPI_Database-Public.csv:
```diff
@@ -0,0 +1,3 @@
+cpiId,status,publicKeyIdentifier 
+ABCdef123,ACTIVE,
+123456,ACTIVE,https://winnf.memberclicks.net/mcdatafiles/receiptattach/winnf/12764702/8960734/Public_Key_Dummy_File.txt
```
The CPI database file is provided by the organization with authority which will be WinnForum in this case. I think we shall keep what we have from the WinnForum. Please advise if you think otherwise

jni2000 commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -174,11 +194,17 @@ def generate_WDB_2_default_config(self, filename):
+        'port': 8003,
+        'fileUrl': '/allsitedata',
+        'filePath': os.path.join('testcases', 'testdata', 'wdb_2', 'CPI_Database-Public.csv')
+    }
```
Is this related to the multiple CPI file comments ? If so, we will address the needs when that one is clarified.

jni2000 commented:
Taran

We have addressed most of your review comments except the followings
1. the private key file issue -- we are working on it
2. the PAL record format -- please can you share with me the format you have ? I believe we used the latest from Kate.
3. the three comments related to supporting multiple CPI files, I do not recall we have such a requirements and I am calcifying this with Kate right now. 

Please take a look of others and let's move on. Thanks.

kate-harrison commented:
2. Please use the format here: https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf

3. Yes, this is a requirement.

jni2000 commented:
2. Just double checked, the pal record file we used is same as the one in https://workspace.winnforum.org/higherlogic/ws/public/download/4488/WINNF-TS-0245-V1.0.0%20PAL%20Database%20.pdf.
3. Is there a specific reason that we need to treat this different than the PAL database file we discussed recently ? I would propose we keep them consistent.

kate-harrison commented on src/harness/testcases/testdata/wdb_2/WDB_2_CPI_Private_Key.txt:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Confirmed, only the public key should be distributed. We don't need to even store the private key in the repo.

kate-harrison commented:
2. It appears as though the example in that file is out of date. Please refer to the end of page 5: "Date of licensee FCC registration. It is expressed using the format, YYYYMM-DDThh:mm:ssZ, as defined by [N.6]."

3. Yes, we need to serve the public key files in addition to the index, whereas the PAL DB has only one file it needs to serve.

jni2000 commented:
Thanks, Kate. Fixed the PAL record already. For helping us speed up, please can you see if my understanding about the multiple CPI files is right:

1. the CPI file will be of the format as defined in the file from the WinnF link
2, we will have multiple of them and the test codes just need to put them in a indexed order for SAS UUT to retrieve.

TaranFurnell commented:
See the comment I just wrote, the gist is a there is the index file which the SAS will be informed about. Then each CPI public key that is referred to by that index file must also be provided by this server.

jni2000 commented:
what will be the index file format ? Is there a document on this ? Will it be like a table of rows of the format like "index, cpiFileName" ?

jni2000 commented:
Taran and Kate, we have addressed all the review comments and made necessary changes, please review.

Some special notes:
1. In the CPI test case, we need both the CPI public key and private key, but for different purposes
2. the public key is part of the database data to be retrieved by SAS UUT for decrypting the CPI signed data and thus put into the database data folder "/src/harness/testcases/testdata/cpi_db"
3. the private key is used for the test harness codes to encrypting the CPI data when mimicking the CBSD with CPI signed data, this is test data and thus put into "/src/harness/testcases/testdata/"

This is what we believe the right arrangement, please advise if you think otherwise.

Thanks.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -92,13 +92,20 @@ def generate_WDB_1_default_config(self, filename):
 
     conditionals = [conditionals_b]
 
```
I think it makes the most sense to save the PAL records above to a file, then use that file in the PAL DB config. Otherwise it's not clear that the CBSDs and the PAL records will be self-consistent (same user ID, right license area, etc.). I suggest that you take the PAL records created above, save them to a file during the default config generation (same format as the file you already have), then give that filename as part of the default config.

The same applies to the CPI information, since you will only need the private keys when creating the config (and not during the test case itself).

janetliuboston replied:
Updated as per suggestion.
Thanks for the suggestion. This is a good way of making CBSDs and PAL records consistent and use the same in the database.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +233,38 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
     # responseCode should be 103 (INVALID_VALUE).
     for registration_response in registration_responses:
-      self.assertEqual(registration_response['response']['responseCode'], 103)
+      self.assertEqual(registration_response['response']['responseCode'], 130)
```
Please revert.

janetliuboston replied:
Reverted.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +233,38 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+    cpi_database.start()
+
+    # Prepare the file list.
+    files = {('/' + file['public_key_file']): file['public_key_file'] for file in config['cpiDatabaseConfig']['cpis']}
```
Please don't use the var name `file` since it is a Python keyword. Probably `cpi` makes more sense here anyway.

janetliuboston replied:
Fixed as per suggestion.

janetliuboston commented:
Hi, This is ready for review again. Thanks!

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -92,13 +93,26 @@ def generate_WDB_1_default_config(self, filename):
+
+    # Create the pal_db record file consistent with CBSDs and PAL records.
+    with open(os.path.join(pal_db_relative_file_path), 'w+') as file_handle:
+      file_handle.write(json.dumps(pal_records_a,pal_records_b,indent=2))
```
As written, this will write a list of lists to the JSON file. Instead, we want just one list of PAL DB records.

janetliuboston replied:
We have doubled checked it. This statement 
json.dumps(pal_records_a,pal_records_b,indent=2)
write a list of PAL records.

kate-harrison replied:
Are you sure that `pal_records_b` is written out? I did an experiment locally and it doesn't seem to be the case.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -92,13 +93,26 @@ def generate_WDB_1_default_config(self, filename):
+    pal_db_relative_file_path = os.path.join('testcases', 'testdata', 'pal_db', 'pal_db_record.json')
+
+    # Create the pal_db record file consistent with CBSDs and PAL records.
+    with open(os.path.join(pal_db_relative_file_path), 'w+') as file_handle:
```
Why do you need another call to os.path.join() here?

janetliuboston replied:
Not required, hence removed.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -160,11 +187,11 @@ def generate_WDB_2_default_config(self, filename):
 
     # Read private keys for the CPI users
-    with open(os.path.join('testcases', 'testdata', 'wdb_2', 'WDB_2_CPI_Private_Key.txt'),
+    with open(os.path.join('testcases', 'testdata','WDB_2_CPI_Private_Key.txt'),
```
(1) You should be generating the public+private key pair in this file (see the REG test cases for an example, it's a single function call), not reading them from disk.
(2) You need write out only the public key, NOT the private key.
(3) Line 194 uses the same file. The CPIs need to have distinct keys.
(4) As you did above, please save the filename in a local var and then use that directly in the config.

janetliuboston replied:
Fixed as per suggestion.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -174,11 +201,25 @@ def generate_WDB_2_default_config(self, filename):
+            'status': 'ACTIVE',
+            'publicKeyIdentifierFile': os.path.join('testcases', 'testdata', 'cpi_db', 'CPI_Database-Public.csv')},
+            {
+                'cpiId': cpi_id_d,
```
Please use consistent indentation for these two CPI dicts so that it's clear they're in the same list.

janetliuboston replied:
Fixed as per suggestion.

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+    config_keys = config['cpiDatabaseConfig']['cpis'][0].keys()
+
+    with open(config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'], 'w+') as output_file:
+        dict_writer = csv.DictWriter(output_file, config_keys)
```
This will not write them in a deterministic order, since line 265 is not deterministic. (Moreover, the dict keys are not the same as the column names.) Please hard-code the column names.

janetliuboston replied:
Fixed as per suggestion

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+    with open(config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'], 'w+') as output_file:
+        dict_writer = csv.DictWriter(output_file, config_keys)
+        dict_writer.writeheader()
+        dict_writer.writerows(config['cpiDatabaseConfig']['cpis'])
```
This will not produce full URLs for the PKI file, which is needed.

janetliuboston replied:
Fixed as per suggestion

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+    files = {('/' + cpi['publicKeyIdentifierFile']):
+                 cpi['publicKeyIdentifierFile'] for cpi in config['cpiDatabaseConfig']['cpis']}
+    files[config['cpiDatabaseConfig']['indexUrl']] = \
+        os.path.join('testcases', 'testdata', 'cpi_db', 'indexUrl.csv')
```
Let's call it just "index.csv".

janetliuboston replied:
Fixed as per suggestion

kate-harrison commented on src/harness/testcases/WINNF_FT_S_WDB_testcase.py:
```diff
@@ -205,24 +246,48 @@ def test_WINNF_FT_S_WDB_2(self, config_filename):
+    # Generate the CSV file and write the content of it.
+    config_keys = config['cpiDatabaseConfig']['cpis'][0].keys()
+
+    with open(config['cpiDatabaseConfig']['cpis'][0]['publicKeyIdentifierFile'], 'w+') as output_file:
```
This is the wrong file. You should be writing index.csv (which is then used on line 275), not overwriting the PKI files.

Please store the path in a local var and use that var below, to reduce the chance for error.

Please call the file index.csv.

kate-harrison commented on src/harness/testcases/testdata/WDB_2_CPI_Private_Key.txt:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Not needed.

janetliuboston replied:
WDB_2_CPI_Private_Key.txt is removed

kate-harrison commented on src/harness/testcases/testdata/cpi_db/CPI_Database-Public.csv:
```diff
@@ -0,0 +1,2 @@
+cpiId,status,publicKeyIdentifier 
```
Why is this here?

janetliuboston replied:
WDB_2_CPI_Private_Key.txt is removed

kate-harrison commented on src/harness/testcases/testdata/cpi_db/WDB_2_CPI_Private_Key.txt:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Not needed.

kate-harrison commented on src/harness/testcases/testdata/pal_db/pal_db_record.json:
```diff
@@ -0,0 +1,30 @@
+{
```
This should be a list of PAL DB records (i.e. there should be `[]` on the outermost level). Please ensure that the file generated by the default config follows that format, but please do not commit that file.

janetliuboston replied:
We have doubled checked it. This statement 
json.dumps(pal_records_a,pal_records_b,indent=2)
write a list of PAL records.

janetliuboston commented:
Updated with the latest. Please review. Thanks!

kate-harrison commented:
I think we can close this now, yes?

janetliuboston commented:
Yes, this is superseded by https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/pull/599
